File: Quick_Sort/Haskell/NorfairKing/Quicksort.hs
module Quicksort ( quicksort ) where

quicksort [] = []
quicksort (pivot:rest) = (quicksort smaller) ++ [pivot] ++ (quicksort greater)
    where
        greater = [ x | x <- rest, x >  pivot ]
        smaller = [ x | x <- rest, x <= pivot ]

File: Quick_Sort/Go/jcla1/quick_sort.go
package quickSort

import "math/rand"

func QuickSort(arr []int) []int {
	if len(arr) < 2 {
		return arr
	}

	pivotIndex := rand.Intn(len(arr))
	pivot := arr[pivotIndex]

	arr = append(arr[:pivotIndex], arr[pivotIndex+1:]...)
	less, greater := make([]int, 0), make([]int, 0)

	for _, v := range arr {
		if v < pivot {
			less = append(less, v)
		} else {
			greater = append(greater, v)
		}
	}

	// Merge the recursively sorted slices back together
	return append(append(QuickSort(less), pivot), QuickSort(greater)...)
}
